### PR TITLE
*: export implicit rowid and use it in CRUD

### PR DESCRIPTION
**Pull request by coocood**
Use a special identier '_tidb_rowid' for implicit rowid.
Makes tables without primary key or unique key easier to debug and replicate by binlog.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @iamxy @GregoryIan PTAL

File: executor/rowid_test.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/pingcap/tidb/util/testkit"
+)
+
+func (s *testSuite) TestExportRowID(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int, b int)")
+	tk.MustExec("insert t values (1, 7), (1, 8), (1, 9)")
+	tk.MustQuery("select *, _tidb_rowid from t").
```
Comment (holys) [inline executor/rowid_test.go]: Should we support syntax like ` select _tidb_rowid, * from no_pk;`?

File: executor/write.go [hunk 2]
```diff
@@ -1181,14 +1183,19 @@ func (e *InsertValues) getColumns(tableCols []*table.Column) ([]*table.Column, e
...
 		if err != nil {
 			return nil, errors.Errorf("INSERT INTO %s: %s", e.Table.Meta().Name.O, err)
 		}
 	} else {
 		// If e.Columns are empty, use all columns instead.
 		cols = tableCols
 	}
+	for _, col := range cols {
+		if col.Name.L == model.ExtraHandleName.L {
+			e.hasExtraHandle = true
```
Comment (ngaut) [inline executor/write.go]: We should break the loop after we have extra internal handle.

File: executor/rowid_test.go (hunk 1 above)
Comment (coocood) [reply executor/rowid_test.go]: No, MySQL doesn't support this syntax, wildcard must be the first field.

**Comment by holys** [comment]
Is that an elegant way to know whether a specific version of tidb supports explicit `_tidb_rowid` or not?

**Comment by coocood** [comment]
@holys 
Only tables don't have integer primary key have _tidb_rowid, we have to check it for a specific table.
I think `select _tidb_rowid from t limit 0` is ok.

**Comment by holys** [comment]
@coocood   OK.  I use `select _tidb_rowid from t limit 0` to check it.

File: table/column.go [hunk 3]
```diff
@@ -77,12 +77,17 @@ func ToColumn(col *model.ColumnInfo) *Column {
...
+func FindCols(cols []*Column, names []string, pkIsHandle bool) ([]*Column, error) {
 	var rcols []*Column
 	for _, name := range names {
 		col := FindCol(cols, name)
 		if col != nil {
 			rcols = append(rcols, col)
+		} else if name == model.ExtraHandleName.L && !pkIsHandle {
+			col := &Column{}
+			col.ColumnInfo = model.NewExtraHandleColInfo()
+			col.ColumnInfo.Offset = len(cols)
```
Comment (WangXiangUSTC) [inline table/column.go]: is the implicit column  always the last column? if user add a new column, is the implicit column's position be changed?

File: table/tables/tables.go [hunk 4]
```diff
@@ -555,11 +568,21 @@ func (t *Table) RemoveRecord(ctx sessionctx.Context, h int64, r []types.Datum) e
...
-		colIDs := make([]int64, 0, len(t.Cols()))
-		for _, col := range t.Cols() {
+		cols := t.Cols()
+		colIDs := make([]int64, 0, len(cols)+1)
+		for _, col := range cols {
 			colIDs = append(colIDs, col.ID)
 		}
-		err = t.addDeleteBinlog(ctx, r, colIDs)
+		var binlogRow []types.Datum
+		if !t.meta.PKIsHandle {
```
Comment (GregoryIan) [inline table/tables/tables.go]: how about unifying the processing methods, with pk or not?

File: table/tables/tables.go [hunk 5]
```diff
@@ -344,11 +350,18 @@ func (t *Table) getRollbackableMemStore(ctx sessionctx.Context) kv.RetrieverMuta
 // AddRecord implements table.Table AddRecord interface.
 func (t *Table) AddRecord(ctx sessionctx.Context, r []types.Datum, skipHandleCheck bool) (recordID int64, err error) {
 	var hasRecordID bool
-	for _, col := range t.Cols() {
-		if col.IsPKHandleColumn(t.meta) {
-			recordID = r[col.Offset].GetInt64()
-			hasRecordID = true
-			break
+	cols := t.Cols()
+	if len(r) > len(cols) {
```
Comment (GregoryIan) [inline table/tables/tables.go]: did already check recordID and pk wether they are same?

File: table/column.go (hunk 3 above)
Comment (coocood) [reply table/column.go]: Yes, the offset is calculated at runtime.

**Comment by WangXiangUSTC** [comment]
if table already had unique key, it seems no need add implicit column @coocood

**Comment by coocood** [comment]
@WangXiangUSTC 
Use the recordID to update or delete will be more efficient, it avoids an index lookup.

**Comment by WangXiangUSTC** [comment]
LGTM

**Comment by coocood** [comment]
@shenli PTAL

File: table/column.go [hunk 6]
```diff
@@ -77,12 +77,17 @@ func ToColumn(col *model.ColumnInfo) *Column {
 }
 
 // FindCols finds columns in cols by names.
-func FindCols(cols []*Column, names []string) ([]*Column, error) {
+func FindCols(cols []*Column, names []string, pkIsHandle bool) ([]*Column, error) {
```
Comment (shenli) [inline table/column.go]: Add a comment for the newly added argument.

File: table/column.go [hunk 7]
```diff
@@ -77,12 +77,17 @@ func ToColumn(col *model.ColumnInfo) *Column {
...
 	var rcols []*Column
 	for _, name := range names {
 		col := FindCol(cols, name)
 		if col != nil {
 			rcols = append(rcols, col)
+		} else if name == model.ExtraHandleName.L && !pkIsHandle {
```
Comment (shenli) [inline table/column.go]: Could we new a Column object in the Table meta? So we could avoid creating new object here.

Comment (coocood) [reply table/column.go]: This is not always created, only if the user explicitly specify `_tidb_rowid`.

**Comment by coocood** [comment]
@shenli PTAL

File: table/tables/tables.go [hunk 8]
```diff
@@ -228,15 +228,16 @@ func (t *Table) UpdateRecord(ctx sessionctx.Context, h int64, oldData, newData [
 	if err != nil {
 		return errors.Trace(err)
 	}
+	numCols := len(newData) + 1
```
Comment (zimulala) [inline table/tables/tables.go]: Please add comments about why we need to add 1. 
I'm afraid we will forget why.

**Comment by coocood** [comment]
@zimulala PTAL

**Comment by shenli** [comment]
LGTM

**Pull request merged**